Add documentation for SSO (#2940)
opensearch-project/documentation-website: _observing-your-data/sso.md

---
layout: default
title: Simple Schema for Observability
nav_order: 120
---

# Simple Schema for Observability
Introduced 2.6
{: .label .label-purple }

Starting with OpenSearch 2.6, a standardization for conforming to a common and unified observability schema has been introduced: Simple Schema for Observability (SSO). [Observability]({{site.url}}{{site.baseurl}}/observing-your-data/index/) is a collection of plugins and applications that let you visualize data-driven events by using Piped Processing Language to explore, discover, and query data stored in OpenSearch.

> With the schema in place, observability tools can ingest, automatically extract, aggregate data, and create custom common dashboards, making it easier to understand the system at a higher level.

Author: yeah it is, thanks for catching that. I'll update it globally.

> With the schema in place, observability tools can ingest and automatically extract and aggregate data and create custom common dashboards making it easier to understand the system at a higher level.

Reviewer: ```suggestion
With the schema in place, observability tools can ingest, automatically extract, aggregate data, and create custom common dashboards making it easier to understand the system at a higher level.
```

> SSO is inspired by both [OTEL](https://opentelemetry.io/docs/) and Elastic Common Schema (ECS) and uses a variety of events that ECS has to offer for logs with the OTEL metadata.

Author: I simplified this sentence and also linked to the Amazon docs on ECS events as it's a pretty comprehensive and detailed guide.

Alerts will be supported in a future release.
{: .note }

### Use cases

Use cases for SSO include:

> - Ingest observability data from different data types.

Reviewer 2: Revise bullets to make them parallel. Suggested revisions:

- Ingesting...
- Moving from non-transferable, proprietary configurations to consolidated, sharable solutions that allow users...and have capabilities to display data analyses.
- Conforming Data Prepper and SSO for metrics, with gradual support of traces and logs.

* Move from proprietary configurations that are non-transferable to a consolidated sharable Observability solution that allows users to ingest any type of telemetry data from any type of provider and have the capability of displaying an analysis of this data.
* Data Prepper conforms to the SSO schema for metrics and will gradually support traces and logs.

> [Data Prepper's]({{site.url}}{{site.baseurl}}/data-prepper/index/) existing traces mapping has enriched data in the form of serviceMap, which differs from SSO traces. This will consolidate with Observability by using the SSO traces schema and introduce serviceMap as an enriched calculated field.

Author: "serviceMap" came from the tech review, however it looks like in Data Prepper's documentation for trace analytics, the configurable parameter is "service-map." I have updated the paragraph to reflect that, as well as linking to relevant documentation. I have re-written the sentences to be more clear. Let me know if that works in your second pass. Thanks.

> [Data Prepper's]({{site.url}}{{site.baseurl}}/data-prepper/index/) existing traces mapping has enriched data in the form of serviceMap which differs from SSO traces. In the future this will consolidate with Observability by using the SSO traces schema and introduce serviceMap as an enriched calculated field.

Reviewer 2: Insert comma after "serviceMap." Suggested revision: "**This will**...."

{: .note }

> ## Traces and Metrics

Reviewer 2: Lowercase "Metrics"

> Schema definitions for traces and metrics are defined and supported by Observability plugin.

Reviewer 2: "...supported by **the**..."

> This includes:

Reviewer 2: suggested revision: "These schema definitions include: "

- Index structure (mapping).
- [Index naming conventions](https://github.com/opensearch-project/observability/issues/1405).
- JSON schema for enforcement and validation of the structure.

> - [Integration](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3412) feature for adding pre-canned dashboards and assets.

Reviewer 3: It looks like we're missing articles (a/an/the) in most of the list items. Remember that the list items should complete the stem if the stem is a fragment. Also, let's use a more formal/precise word than "pre-canned," which may not be understood by all readers.

Author: Unfragmented the bullet points and used more formal wording. Thanks!

---
layout: default
title: Simple Schema for Observability
nav_order: 120
---

# Simple Schema for Observability
Introduced 2.6
{: .label .label-purple }

OpenSearch 2.6 introduced a standardization for conforming to a common and unified observability schema: Simple Schema for Observability (SSO). [Observability]({{site.url}}{{site.baseurl}}/observing-your-data/index/) is a collection of plugins and applications that let you visualize data-driven events by using Piped Processing Language (PPL) to explore and query data stored in OpenSearch.

With the schema in place, Observability tools can ingest, automatically extract, and aggregate data and create custom dashboards, making it easier to understand the system at a higher level.

> SSO is inspired by both [OpenTelemetry](https://opentelemetry.io/docs/) and the Elastic Common Schema (ECS) and uses [ECS events](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs_cwe_events.html) for logs with the OpenTelemetry metadata.

Reviewer 3: Unfortunately I'm still not following "uses ECS events for logs with the OpenTelemetry metadata." It also looks like we define ECS as Elastic Common Schema but then refer to ECS events, in which case ECS stands for Amazon Elastic Container Service (Amazon ECS). Let's rewrite this sentence so that ECS only has one meaning and so that we are only providing necessary information (does the reader really need to know what SSO was inspired by?).

Author: Re-wrote as per our conversation.

> SSO is compatible with the [OpenTelemetry](https://opentelemetry.io/docs/) (OTEL) schema including traces, logs, and metrics. Alerts are also supported. SSO also accepts Elastic Common Schema (ECS) logs when structured logs such as these are present.

Reviewer 4: Alerts will be supported in the future

> SSO is compatible with the [OpenTelemetry](https://opentelemetry.io/docs/) (OTEL) schema including traces, logs, and metrics. Alerts are also supported. SSO also accepts Elastic Common Schema (ECS) logs when structured logs such as these are present.

Reviewer 4: SSO schema is inspired by both OTEL & ECS and tries to use the large variety ECS has to offer for logs with the OTEL metadata

> SSO is inspired by both [OpenTelemetry](https://opentelemetry.io/docs/) and the Elastic Common Schema (ECS) and uses Amazon Elastic Container Service ([Amazon ECS](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs_cwe_events.html)) event logs and OpenTelemetry metadata.

Reviewer 5: why are we talking about amazon ecs

Alerts will be supported in a future release.
{: .note }

## Use cases

Use cases for SSO include:

* Ingesting Observability data from different data types.

> * Moving from proprietary configurations that are non-transferable to a consolidated, sharable Observability solution that allows users to ingest and display an analysis of any type of telemetry data from any type of provider.

Reviewer 3: In this bullet and the next one, confirm that the capitalized instances of "Observability" are actually references to the proper noun and not just generic references to the concept of observability, which should be lowercase.

> - Move away from a dependence on the Data Prepper plugin.

Reviewer 4: please remove this phrase since its not correct - data-prepper is conforming to SSO schema for metrics and will gradually conform so with traces and logs

## Data Prepper

> SSO allows for independent Observability ingestion that has been handled by the [Data Prepper]({{site.url}}{{site.baseurl}}/data-prepper/index/) plugin up until now by configuring an [OTel trace source](https://github.com/opensearch-project/data-prepper/tree/main/data-prepper-plugins/otel-trace-source).

Reviewer 4: data-prepper existing traces mapping have enriched data in the form of serviceMap whic differs from sso traces. In the future this will consolidate with Observability by using the sso traces schema and introducing the serviceMap as an enriched calculated field

> Data Prepper's [trace mapping]({{site.url}}{{site.baseurl}}/data-prepper/common-use-cases/trace-analytics/) currently provides `service-map` data in a different way than SSO traces. To make the trace mapping compatible with Observability, it will be integrated with the SSO traces schema and will introduce `service-map` as an enriched field.

Reviewer 4: in a different way than SSO traces. -> in addition and thus enriching the standard trace signal with additional calculated field.

## Traces and metrics

Schema definitions for traces and metrics are defined and supported by the Observability plugin.

These schema definitions include:

> - The Index structure (mapping).

Reviewer 3: ```suggestion
- The index structure (mapping).
```

- Index naming conventions.
- JSON schema for enforcement and validation of the structure.

>  

Reviewer 4: - Index naming convention see https://github.com/opensearch-project/observability/issues/1405
 - Integration feature for adding pre-canned dashboards and assets https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3412

> - The [Index naming conventions](https://github.com/opensearch-project/observability/issues/1405).

Reviewer 3: ```suggestion
- The [index naming conventions](https://github.com/opensearch-project/observability/issues/1405).
```

> - The [Integration](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3412) feature for adding pre-configured dashboards and assets.

Reviewer 3: ```suggestion
- The [integration](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3412) feature for adding preconfigured dashboards and assets.
```
